# Проект по SQL: Анализ базы данных.

***Проект выполнен Вяжевым Даниилом Юрьевичом от 01.07.2022 на платформе Яндекс Практикум.***

# Описание проекта:

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.


# Цели исследования:

Проанализировать базу данных для того чтобы сформулировать ценностное предложение для нового продукта
Изучить информацию о книгах, издательствах, авторах и пользовательских обзоров книг.
Описание данных

**Таблица `books`**

Содержит данные о книгах:

`book_id` — идентификатор книги;

`author_id` — идентификатор автора;

`title` — название книги;

`num_pages` — количество страниц;

`publication_date` — дата публикации книги;

`publisher_id` — идентификатор издателя.


**Таблица `authors`**

Содержит данные об авторах:

`author_id` — идентификатор автора;

`author` — имя автора.


**Таблица `publishers`**


Содержит данные об издательствах:

`publisher_id` — идентификатор издательства;

`publisher` — название издательства;


**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

`rating_id` — идентификатор оценки;

`book_id` — идентификатор книги;

`username` — имя пользователя, оставившего оценку;

`rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

`review_id` — идентификатор обзора;

`book_id `— идентификатор книги;

`username` — имя пользователя, написавшего обзор;

`text `— текст обзора.

# Содержание:
<a id='content'></a>
[1 Шаг. Импорт библиотек](#part_1)

[2 Шаг. Изучение общей информации](#part_2)

[3 Шаг. Задачи исследования](#part_3)

[3.1  Посчитайте, сколько книг вышло после 1 января 2000 года;](#part_3.1)

[3.2  Для каждой книги посчитайте количество обзоров и среднюю оценку;](#part_3.2)

[3.3  Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;](#part_3.3)

[3.4  Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;](#part_3.4)

[3.5  Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.](#part_3.5)

[Вывод](#part_4)

## Шаг. Импорт библиотек 
<a id='part_1'></a>

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student',
             'pwd': 'Sdf4$2;d-d30pp', 
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 'port': 6432,
             'db': 'data-analyst-final-project-db'} 
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],  db_config['pwd'],  
                                                         db_config['host'],  db_config['port'],  
                                                         db_config['db'])

In [3]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

##  Шаг. Изучение общей информации 
<a id='part_2'></a>

In [4]:
books = '''SELECT * FROM books LIMIT 5'''
pd.io.sql.read_sql(books, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
authors = '''SELECT * FROM authors LIMIT 5'''
pd.io.sql.read_sql(authors, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
ratings = '''SELECT * FROM ratings LIMIT 5'''
pd.io.sql.read_sql(ratings, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
reviews = '''SELECT * FROM reviews LIMIT 5'''
pd.io.sql.read_sql(reviews, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [8]:
publishers = '''SELECT * FROM publishers LIMIT 5'''
pd.io.sql.read_sql(publishers, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Получили 5 таблиц: `books, authors, ratings, reviews и publishers`. В каждой из таблиц содержится информация о книгах, издательствах, авторах и пользовательских обзоров книг. 

## Шаг. Задачи исследования
<a id='part_3'></a>

### Посчитайте, сколько книг вышло после 1 января 2000 года;
<a id='part_3.1'></a>

In [9]:
task_1 = '''
SELECT COUNT(publication_date)
FROM books
WHERE publication_date >= '2000-01-01'
'''
books_released = pd.io.sql.read_sql(task_1, con = engine)
print('Количество книг, вышедших после 1 января 2000 года: {:.0f}'.format(books_released.iloc[0][0]))

Количество книг, вышедших после 1 января 2000 года: 821


### Для каждой книги посчитайте количество обзоров и среднюю оценку;
<a id='part_3.2'></a>

Рассчитали топ-5 по средней оценке.

In [10]:

query = '''

  SELECT books.book_id,
         books.title, 
         ratings.avg_rating,
         reviews.reviews_count
  FROM books
  LEFT JOIN (SELECT book_id,
                    AVG(rating) as avg_rating
             FROM ratings
             GROUP BY book_id) as ratings ON books.book_id = ratings.book_id
  LEFT JOIN (SELECT book_id,
                    COUNT (review_id) as reviews_count
                    FROM reviews
                    GROUP BY book_id) as reviews ON books.book_id = reviews.book_id
  ORDER BY ratings.avg_rating DESC
  '''

book_count_mean = pd.io.sql.read_sql(query, con = engine)
display(book_count_mean)

book_id                                              title  avg_rating  \
0        518               Pop Goes the Weasel (Alex Cross  #5)        5.00   
1        732  The Ghost Map: The Story of London's Most Terr...        5.00   
2        347  In the Hand of the Goddess (Song of the Liones...        5.00   
3        610                           Tai-Pan (Asian Saga  #2)        5.00   
4        330  How to Be a Domestic Goddess: Baking and the A...        5.00   
..       ...                                                ...         ...   
995      915  The World Is Flat: A Brief History of the Twen...        2.25   
996      371                                              Junky        2.00   
997      316                  His Excellency: George Washington        2.00   
998      202                                      Drowning Ruth        2.00   
999      303                               Harvesting the Heart        1.50   

     reviews_count  
0              2.0  
1              2.0  
2              2.0  
3              2.0  
4              1.0  
..             ...  
995            3.0  
996            2.0  
997            2.0  
998            3.0  
999            2.0  

[1000 rows x 4 columns]

In [11]:
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  avg_rating  \
0        518               Pop Goes the Weasel (Alex Cross  #5)        5.00   
1        732  The Ghost Map: The Story of London's Most Terr...        5.00   
2        347  In the Hand of the Goddess (Song of the Liones...        5.00   
3        610                           Tai-Pan (Asian Saga  #2)        5.00   
4        330  How to Be a Domestic Goddess: Baking and the A...        5.00   
..       ...                                                ...         ...   
995      915  The World Is Flat: A Brief History of the Twen...        2.25   
996      371                                              Junky        2.00   
997      316                  His Excellency: George Washington        2.00   
998      202                                      Drowning Ruth        2.00   
999      303                               Harvesting the Heart        1.50   

     reviews_count  
0              2.0  
1              2.0  
2              2.0  
3              2.0  
4              1.0  
..             ...  
995            3.0  
996            2.0  
997            2.0  
998            3.0  
999            2.0  

[1000 rows x 4 columns]

Наибольшее количество обзоров вышло по книге Twilight. При этом рейтинг у этой книги ниже остальных.

Всего в данных присутсвует 1000 книг.

Рассмотрим вопрос более объективно и рассчитаем топ-5 по кол-ву обзоров.

In [12]:
query = '''

  SELECT books.book_id,
         books.title, 
         ratings.avg_rating,
         reviews.reviews_count
  FROM books
  LEFT JOIN (SELECT book_id,
                    AVG(rating) as avg_rating
             FROM ratings
             GROUP BY book_id) as ratings ON books.book_id = ratings.book_id
  LEFT JOIN (SELECT book_id,
                    COUNT (review_id) as reviews_count
                    FROM reviews
                    GROUP BY book_id) as reviews ON books.book_id = reviews.book_id
  WHERE reviews.reviews_count != 'NaN'::NUMERIC
  ORDER BY reviews.reviews_count DESC
  '''

In [13]:
display(pd.io.sql.read_sql(query, con = engine).head())
display(pd.io.sql.read_sql(query, con = engine).count())

book_id                                   title  avg_rating  reviews_count
0      948                 Twilight (Twilight  #1)    3.662500              7
1      656                          The Book Thief    4.264151              6
2      497               Outlander (Outlander  #1)    4.125000              6
3      963                     Water for Elephants    3.977273              6
4      696  The Da Vinci Code (Robert Langdon  #2)    3.830508              6

book_id          994
title            994
avg_rating       994
reviews_count    994
dtype: int64

А здесь 995 книг, 5 книг где-то были потеряны.

И рассмотрим наши книги-потеряшки, которые остались без ревью:

In [14]:
query = '''

  SELECT count(review_id) AS cnt_review,
      COUNT(b.book_id) cnt,
      b.title AS book
  FROM books b
  LEFT JOIN reviews rv ON b.book_id = rv.book_id
  WHERE review_id IS null
  GROUP BY book
  ORDER BY 2 DESC
  '''
lost = pd.io.sql.read_sql(query, con = engine)
display( lost.head())

cnt_review  cnt                                               book
0           0    1    Anne Rice's The Vampire Lestat: A Graphic Novel
1           0    1  Disney's Beauty and the Beast (A Little Golden...
2           0    1                          Essential Tales and Poems
3           0    1                               Leonardo's Notebooks
4           0    1   The Cat in the Hat and Other Dr. Seuss Favorites

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
<a id='part_3.3'></a>

In [15]:
task_3 = '''
SELECT publisher AS publisher_name,
    COUNT (books.book_id) AS cnt_books
FROM books 
    INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP  BY publisher
ORDER  BY cnt_books DESC
'''

publisher_most_books = pd.io.sql.read_sql(task_3, con = engine)
publisher_most_books
print('Издательство', publisher_most_books.iloc[0, 0], 'выпустило наибольшее число книг -', 
      publisher_most_books.iloc[0, 1], 'шт. толще 50 страниц.'  )

Издательство Penguin Books выпустило наибольшее число книг - 42 шт. толще 50 страниц.


### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
<a id='part_3.4'></a>

In [16]:
task_4 = '''
SELECT authors.author AS author_name,
AVG(r.average_rating) as avg_rating
FROM books 
INNER JOIN (SELECT book_id, COUNT(rating) AS cnt,
             AVG(rating) AS average_rating
      FROM ratings
      GROUP BY book_id) AS r
ON r.book_id = books.book_id
INNER JOIN authors ON authors.author_id = books.author_id
WHERE r.cnt >= 50
GROUP BY author_name
ORDER BY avg_rating DESC
LIMIT 1
'''

auth_books_avg_rating = pd.io.sql.read_sql(task_4, con = engine)
print('Автор -', auth_books_avg_rating.iloc[0,0], 'с самой высокой средней оценкой книг:',
      auth_books_avg_rating.iloc[0,1] )

Автор - J.K. Rowling/Mary GrandPré с самой высокой средней оценкой книг: 4.283844038868471


### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.
<a id='part_3.5'></a>

In [17]:
task_5 = '''
SELECT
    AVG(sub.cnt_text)
FROM
(SELECT 
   reviews.username,
   COUNT(DISTINCT reviews.text) AS cnt_text,
   COUNT(DISTINCT ratings.book_id) AS cnt_book
FROM reviews
     INNER JOIN ratings ON ratings.username = reviews.username
GROUP BY
   reviews.username
HAVING
   COUNT(DISTINCT ratings.book_id)>50) AS sub
'''

avg_cnt_reviews = pd.io.sql.read_sql(task_5, con = engine)
avg_cnt_reviews.head(10)
print('Среднее количество обзоров от пользователей, которые поставили больше 50 оценок равно:', 
      avg_cnt_reviews.iloc[0][0].round())

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок равно: 24.0


# Вывод:
<a id='part_4'></a>

* Количество книг, вышедших после 1 января 2000 года: 821;

* Наибольшее количество обзоров вышло по книге Twilight. При этом рейтинг у этой книги ниже остальных;

* Издательство Penguin Books выпустило наибольшее число книг - 42 шт. толще 50 страниц;

* Автор - J.K. Rowling/Mary GrandPré с самой высокой средней оценкой книг: 4.283844038868471;

* Среднее количество обзоров от пользователей, которые поставили больше 50 оценок равно: 24.333333333333332.
